api/payments.js

```diff
@@ -113,89 +80,53 @@ function formatPaymentHelper(account, transaction, callback) {
-                                 _callback);
+  function formatTransactionResponse(message, meta) {
+    if (meta.state === 'validated') {
+      var txJSON = message.tx_json;
+      txJSON.meta = message.metadata;
+      txJSON.validated = message.validated;
+      txJSON.ledger_index = txJSON.inLedger = message.ledger_index;
+      return formatPaymentHelper(payment.source_account, txJSON);
     }
 
-    _callback(null, {
+    return {
       client_resource_id: clientResourceID,
       status_url: urlBase + '/v1/accounts/' + payment.source_account
         + '/payments/' + clientResourceID
-    });
+    };
   }
 
-  function _createPaymentTransaction(remote, _callback) {
-    var transaction;
-    try {
-      transaction = createPaymentTransaction(account, payment);
-    } catch (err) {
-      _callback(err);
-      return;
-    }
-
-    var ledgerIndex;
-    var maxFee = Number(max_fee);
-    var fixedFee = Number(fixed_fee);
+  function prepareTransaction(_transaction, remote) {
+    validate.client_resource_id(clientResourceID);
 
-    if (Number(options.last_ledger_sequence) > 0) {
-      ledgerIndex = Number(options.last_ledger_sequence);
-    } else {
-      ledgerIndex = Number(remote._ledger_current_index)
-        + transactions.DEFAULT_LEDGER_BUFFER;
-    }
-
-    transaction.lastLedger(ledgerIndex);
+    _transaction.lastLedger(Number(options.last_ledger_sequence) ||
+      Number(remote.getLedgerSequence()) + transactions.DEFAULT_LEDGER_BUFFER);
 
-    if (maxFee >= 0) {
-      transaction.maxFee(maxFee);
+    if (Number(options.max_fee) >= 0) {
+      _transaction.maxFee(Number(xrpToDrops(options.max_fee)));
     }
 
-    if (fixedFee >= 0) {
-      transaction.setFixedFee(fixedFee);
+    if (Number(options.fixed_fee) >= 0) {
+      _transaction.setFixedFee(Number(xrpToDrops(options.fixed_fee)));
     }
 
-    transaction.clientID(clientResourceID);
-    _callback(null, transaction);
+    _transaction.clientID(clientResourceID);
+    return _transaction;
   }
 
-  var _options = {
-    validated: options.validated,
+  var _options = _.assign({}, options, {
     clientResourceId: clientResourceID,
-    blockDuplicates: true,
-    saveTransaction: true
-  };
+    blockDuplicates: options.submit !== false,
+    saveTransaction: options.submit !== false
+  });
 
-  var converter = formatTransactionResponse;
-  async.waterfall([
-    _.partial(_createPaymentTransaction, self.remote),
-    function(transaction, _callback) {
-      try {   // for case of missing secret and submit=false not set
-        transact(transaction, self, secret, _options, converter, _callback);
-      } catch (err) {
-        _callback(err);
-      }
-    }
-  ], callback);
+  var initialTx = createPaymentTransaction(account, payment);
+  var transaction = options.submit !== false ? prepareTransaction(
+    initialTx, this.remote) : initialTx;
+  var converter = options.submit !== false ? formatTransactionResponse :
```

Comment:
Duplicating `options.submit !== false` 4 times here. Consider storing the result in a variable